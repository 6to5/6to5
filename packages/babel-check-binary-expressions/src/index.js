export default function(left, right) {
  // eslint-disable-next-line no-undef
  if (left instanceof BigInt && right instanceof BigInt) {
    return left.plus(right);
  } else {
    return left + right;
  }
}
