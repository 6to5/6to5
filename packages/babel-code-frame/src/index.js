import jsTokens, { matchToToken } from "js-tokens";
import esutils from "esutils";
import Chalk from "chalk";

/**
 * Chalk styles for token types.
 */

function getDefs(chalk) {
  return {
    keyword:     chalk.cyan,
    capitalized: chalk.yellow,
    jsx_tag:     chalk.yellow,
    punctuator:  chalk.yellow,
    // bracket:  intentionally omitted.
    number:      chalk.magenta,
    string:      chalk.green,
    regex:       chalk.magenta,
    comment:     chalk.grey,
    invalid:     chalk.white.bgRed.bold,
    gutter:      chalk.grey,
    marker:      chalk.red.bold,
  };
}

/**
 * RegExp to test for newlines in terminal.
 */

const NEWLINE = /\r\n|[\n\r\u2028\u2029]/;

/**
 * RegExp to test for what seems to be a JSX tag name.
 */

const JSX_TAG = /^[a-z][\w-]*$/i;

/**
 * RegExp to test for the three types of brackets.
 */

const BRACKET = /^[()\[\]{}]$/;

/**
 * Get the type of token, specifying punctuator type.
 */

function getTokenType(match) {
  const [offset, text] = match.slice(-2);
  const token = matchToToken(match);

  if (token.type === "name") {
    if (esutils.keyword.isReservedWordES6(token.value)) {
      return "keyword";
    }

    if (
      JSX_TAG.test(token.value) &&
      (text[offset - 1] === "<" || text.substr(offset - 2, 2) == "</")
    ) {
      return "jsx_tag";
    }

    if (token.value[0] !== token.value[0].toLowerCase()) {
      return "capitalized";
    }
  }

  if (token.type === "punctuator" && BRACKET.test(token.value)) {
    return "bracket";
  }

  return token.type;
}

/**
 * Highlight `text`.
 */

function highlight(defs: Object, text: string) {
  return text.replace(jsTokens, function (...args) {
    const type = getTokenType(args);
    const colorize = defs[type];
    if (colorize) {
      return args[0].split(NEWLINE).map((str) => colorize(str)).join("\n");
    } else {
      return args[0];
    }
  });
}

type Position = {
  lineNumber: number,
  colNumber: ?number,
}

/**
 * Create a code frame, adding line numbers, code highlighting, and pointing to a given position.
 */
declare function codeFrame(rawLines: string, start: Position, end: ?Position, opts: Object = {}): string;
declare function codeFrame(rawLines: string, lineNumber: number, colNumber: ?number, opts: Object = {}): string;
export default function codeFrame(rawLines, startOrLineNumber, endOrColNumber, opts) {
  let startLineNumber = undefined;
  let startColNumber = undefined;
  let endLineNumber = undefined;
  let endOrColNumber = undefined;
  let markRange = false;
  if (typeof startOrLineNumber === 'number') {
    startLineNumber = startOrLineNumber;
    if (typeof endOrColNumber === 'number') {
      startColNumber = endOrColNumber;
    }
    console.warn('Passing lineNumber and colNumber is deprecated to babel-code-frame. Pass start position object.');
  } else if (typeof startOrLineNumber === 'object') {
    startLineNumber = startOrLineNumber.lineNumber;
    startColNumber = startOrLineNumber.colNumber;
    if (endOrColNumber !== undefined) {
      if (typeof endOrColNumber === 'object') {
        endLineNumber = endOrColNumber.lineNumber;
        endOrColNumber = endOrColNumber.colNumber;
        markRange = true;
      } else {
        throw new TypeError('end must be an object with lineNumber and colNumber properties');
      }
    }
  } else {
    throw new TypeError('start and end should be objects with lineNumber and colNumber properties');
  }

  startColNumber = Math.max(startColNumber, 0);

  const highlighted = (opts.highlightCode && Chalk.supportsColor) || opts.forceColor;
  let chalk = Chalk;
  if (opts.forceColor) {
    chalk = new Chalk.constructor({ enabled: true });
  }
  const maybeHighlight = (chalkFn, string) => {
    return highlighted ? chalkFn(string) : string;
  };
  const defs = getDefs(chalk);
  if (highlighted) rawLines = highlight(defs, rawLines);

  const linesAbove = opts.linesAbove || 2;
  const linesBelow = opts.linesBelow || 3;

  const lines = rawLines.split(NEWLINE);
  let start = Math.max(startLineNumber - (linesAbove + 1), 0);
  let end   = Math.min(lines.length, startLineNumber + linesBelow);

  if (!startLineNumber && !startColNumber) {
    start = 0;
    end = lines.length;
  }

  const numberMaxWidth = String(end).length;

  const frame = lines.slice(start, end).map((line, index) => {
    const number = start + 1 + index;
    const paddedNumber = ` ${number}`.slice(-numberMaxWidth);
    const gutter = ` ${paddedNumber} | `;
    if (number === startLineNumber) {
      let markerLine = "";
      if (startColNumber) {
        const markerSpacing = line.slice(0, startColNumber - 1).replace(/[^\t]/g, " ");
        markerLine = [
          "\n ",
          maybeHighlight(defs.gutter, gutter.replace(/\d/g, " ")),
          markerSpacing,
          maybeHighlight(defs.marker, "^")
        ].join("");
      }
      return [
        maybeHighlight(defs.marker, ">"),
        maybeHighlight(defs.gutter, gutter),
        line,
        markerLine
      ].join("");
    } else {
      return ` ${maybeHighlight(defs.gutter, gutter)}${line}`;
    }
  }).join("\n");

  if (highlighted) {
    return chalk.reset(frame);
  } else {
    return frame;
  }
}
