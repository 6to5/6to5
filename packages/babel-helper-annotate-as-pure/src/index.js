import * as t from "@babel/types";
export { default as CACHE_KEY } from "./_cache-key";

const PURE_ANNOTATION = "#__PURE__";

const isPureAnnotated = node => {
  const { leadingComments } = node;
  if (leadingComments === undefined) {
    return false;
  }
  return leadingComments.some(comment => /[@#]__PURE__/.test(comment.value));
};

export default function annotateAsPure(pathOrNode) {
  const node = pathOrNode.node || pathOrNode;
  if (isPureAnnotated(node)) {
    return;
  }
  t.addComment(node, "leading", PURE_ANNOTATION);
}
