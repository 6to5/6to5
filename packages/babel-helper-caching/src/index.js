// @flow

import sha256 from "hash.js/lib/hash/sha/256";

type ComputedKey = {
  toString: () => string,
};
type PrimitiveKey = string | number | boolean | null | void;

type KeySet = Array<CacheKey>;
type KeyMap = { [string]: CacheKey };

export type CacheKey = PrimitiveKey | ComputedKey;

export { default as CACHE_KEY } from "./_cache-key";

export default function buildCacheKey(...args: Array<CacheKey>): ComputedKey {
  return {
    inspect: () => {
      return args.length === 1 ? args[0] : args;
    },
    toString: wrap(() => buildObjectHash(args)),
  };
}

buildCacheKey.obj = (arg: KeySet | KeyMap): ComputedKey => {
  return {
    inspect: () => arg,
    toString: wrap(() => buildObjectHash(arg)),
  };
};

buildCacheKey.error = (msg: string, inspectMsg?: string = msg): ComputedKey => {
  return {
    inspect: () => inspectMsg,
    toString: () => {
      throw new Error(msg);
    },
  };
};

buildCacheKey.lazy = (fn: () => CacheKey): ComputedKey => {
  let done = false;
  let value;

  function init() {
    if (done) return;

    value = fn();
    done = true;
  }

  return {
    inspect: () => {
      init();
      return value;
    },
    toString: wrap(() => {
      init();
      return stringify(value);
    }),
  };
};

function wrap(cb: () => string) {
  let generating = false;
  let sha;

  return () => {
    if (sha === undefined) {
      if (generating === true) {
        throw new Error("Cycle detected in cache key graph.");
      }

      generating = true;
      try {
        sha = cb();
      } finally {
        generating = false;
      }
    }
    return sha;
  };
}

function buildObjectHash(obj: KeySet | KeyMap): string {
  const hash = sha256();
  if (Array.isArray(obj)) {
    hash.update("[");
    obj.forEach(item => {
      hash.update(stringify(item));
      hash.update(",");
    });
    hash.update("]");
  } else {
    // simple object, create key from keys and props
    hash.update("{");
    for (const key of Object.keys(obj)) {
      hash.update(key);
      hash.update(":");
      hash.update(stringify(obj[key]));
      hash.update(",");
    }
    hash.update("}");
  }
  return hash.digest("hex");
}

// Validate that a value matches 'CacheKey'.
export function validKey(item: mixed): boolean {
  return (
    item == null ||
    typeof item === "string" ||
    typeof item === "number" ||
    typeof item === "boolean" ||
    (typeof item === "object" &&
      typeof item.toString === "function" &&
      item.toString !== Object.prototype.toString &&
      item.toString !== Array.prototype.toString &&
      item.toString !== Symbol.prototype.toString) ||
    (typeof item === "function" &&
      typeof item.toString === "function" &&
      item.toString !== Function.prototype.toString)
  );
}

function stringify(item: mixed): string {
  if (!validKey(item)) {
    throw new Error(
      "Items must be non-undefined primitives, or have meaningful toString implementations",
    );
  }

  // If we weren't compiling with loose mode, we could do `${item}`
  // $FlowIgnore - Flow doesn't like objects that opt into stringification.
  return "".concat(item);
}
