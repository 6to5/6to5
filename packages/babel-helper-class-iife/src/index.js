import { types as t } from "@babel/core";
import annotateAsPure from "@babel/helper-annotate-as-pure";

const classIIFEs = new WeakMap();

export default function getClassIIFE(path, assumePure) {
  const { node } = path;

  if (classIIFEs.has(node)) {
    return classIIFEs.get(node);
  }

  const iife = t.callExpression(
    t.arrowFunctionExpression([], t.blockStatement([])),
    [],
  );

  classIIFEs.set(node, iife);

  if (assumePure || path.isPure()) {
    annotateAsPure(iife);
  }

  return iife;
}
