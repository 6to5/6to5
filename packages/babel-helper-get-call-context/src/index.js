import type { NodePath } from "@babel/traverse";

export default function(callPath: NodePath): NodePath {
  if (!callPath.isCallExpression() && !callPath.isOptionalCallExpression()) {
    throw new Error(
      `Expected type "CallExpression" or "OptionalCallExpression" ` +
        `but instead got "${callPath.type}".`,
    );
  }

  let calleePath = callPath.get("callee");

  while (
    calleePath.isTSAsExpression() ||
    calleePath.isTypeCastExpression() ||
    calleePath.isTSTypeAssertion() ||
    calleePath.isParenthesizedExpression()
  ) {
    calleePath = calleePath.get("expression");
  }

  return calleePath;
}
