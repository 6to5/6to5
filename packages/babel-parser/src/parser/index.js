// @flow

import type { Options } from "../options";
import type { File, Expression, JSXOpeningElement } from "../types";
import type { PluginList } from "../plugin-utils";
import { getOptions } from "../options";
import StatementParser from "./statement";
import { types as tt } from "../tokenizer/types";

export type PluginsMap = {
  [key: string]: { [option: string]: any },
};

export default class Parser extends StatementParser {
  // Forward-declaration so typescript plugin can override jsx plugin
  +jsxParseOpeningElementAfterName: (
    node: JSXOpeningElement,
  ) => JSXOpeningElement;

  constructor(options: ?Options, input: string) {
    options = getOptions(options);
    super(options, input);

    this.options = options;
    this.inModule = this.options.sourceType === "module";
    this.input = input;
    this.plugins = pluginsMap(this.options.plugins);
    this.filename = options.sourceFilename;
  }

  parse(): File {
    const file = this.startNode();
    const program = this.startNode();
    this.nextToken();
    return this.parseTopLevel(file, program);
  }

  // Convenience method to parse an Expression only
  getExpression(): Expression {
    this.nextToken();
    const expr = this.parseExpression();
    if (!this.match(tt.eof)) {
      this.unexpected();
    }
    expr.comments = this.state.comments;
    return expr;
  }
}

function pluginsMap(plugins: PluginList): PluginsMap {
  const pluginMap: PluginsMap = (Object.create(null): Object);
  for (const plugin of plugins) {
    const [name, options = {}] = Array.isArray(plugin) ? plugin : [plugin, {}];
    if (!pluginMap[name]) pluginMap[name] = options || {};
  }
  return pluginMap;
}
