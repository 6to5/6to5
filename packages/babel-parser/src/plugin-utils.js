// @flow

export type Plugin = string | [string, Object];

export type PluginList = $ReadOnlyArray<Plugin>;

export function hasPlugin(plugins: PluginList, name: string): boolean {
  return plugins.some(plugin => {
    if (Array.isArray(plugin)) {
      return plugin[0] === name;
    } else {
      return plugin === name;
    }
  });
}

export function getPluginOption(
  plugins: PluginList,
  name: string,
  option: string,
) {
  const plugin = plugins.find(plugin => {
    if (Array.isArray(plugin)) {
      return plugin[0] === name;
    } else {
      return plugin === name;
    }
  });

  if (plugin && Array.isArray(plugin)) {
    return plugin[1][option];
  }

  return null;
}

const PIPELINE_PROPOSALS = ["minimal", "smart"];

export function validatePlugins(plugins: PluginList) {
  if (hasPlugin(plugins, "decorators")) {
    if (hasPlugin(plugins, "decorators-legacy")) {
      throw new Error(
        "Cannot use the decorators and decorators-legacy plugin together",
      );
    }

    const decoratorsBeforeExport = getPluginOption(
      plugins,
      "decorators",
      "decoratorsBeforeExport",
    );
    if (decoratorsBeforeExport == null) {
      throw new Error(
        "The 'decorators' plugin requires a 'decoratorsBeforeExport' option," +
          " whose value must be a boolean. If you are migrating from" +
          " Babylon/Babel 6 or want to use the old decorators proposal, you" +
          " should use the 'decorators-legacy' plugin instead of 'decorators'.",
      );
    } else if (typeof decoratorsBeforeExport !== "boolean") {
      throw new Error("'decoratorsBeforeExport' must be a boolean.");
    }
  }

  if (hasPlugin(plugins, "flow") && hasPlugin(plugins, "typescript")) {
    throw new Error("Cannot combine flow and typescript plugins.");
  }

  if (
    hasPlugin(plugins, "pipelineOperator") &&
    !PIPELINE_PROPOSALS.includes(
      getPluginOption(plugins, "pipelineOperator", "proposal"),
    )
  ) {
    throw new Error(
      "'pipelineOperator' requires 'proposal' option whose value should be one of: " +
        PIPELINE_PROPOSALS.map(p => `'${p}'`).join(", "),
    );
  }
}

export const mixinPluginNames: $ReadOnlyArray<string> = [
  "estree",
  "flow",
  "typescript",
  "jsx",
  "placeholders",
];
