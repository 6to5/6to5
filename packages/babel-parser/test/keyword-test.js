import { parse } from "../src";

function run(code, options) {
  return parse(code, { sourceType: "module", ...options });
}
describe("keyword test", function () {
  it("switch if keyword", function () {
    expect(
      run("如果 (true) {}", {
        localizedKeywords: {
          if: "如果",
        },
      }),
    ).toMatchSnapshot();
  });
});
