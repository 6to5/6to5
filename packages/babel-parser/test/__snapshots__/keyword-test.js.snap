// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keyword test switch if keyword 1`] = `
Node {
  "comments": Array [],
  "end": 15,
  "errors": Array [],
  "loc": SourceLocation {
    "end": Position {
      "column": 15,
      "line": 1,
    },
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "program": Node {
    "body": Array [
      Node {
        "alternate": null,
        "consequent": Node {
          "body": Array [],
          "directives": Array [],
          "end": 15,
          "loc": SourceLocation {
            "end": Position {
              "column": 15,
              "line": 1,
            },
            "start": Position {
              "column": 13,
              "line": 1,
            },
          },
          "start": 13,
          "type": "BlockStatement",
        },
        "end": 15,
        "loc": SourceLocation {
          "end": Position {
            "column": 15,
            "line": 1,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "start": 0,
        "test": Node {
          "end": 11,
          "loc": SourceLocation {
            "end": Position {
              "column": 11,
              "line": 1,
            },
            "start": Position {
              "column": 7,
              "line": 1,
            },
          },
          "start": 7,
          "type": "BooleanLiteral",
          "value": true,
        },
        "type": "IfStatement",
      },
    ],
    "directives": Array [],
    "end": 15,
    "interpreter": null,
    "loc": SourceLocation {
      "end": Position {
        "column": 15,
        "line": 1,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "start": 0,
    "type": "Program",
  },
  "start": 0,
  "type": "File",
}
`;
